tests/pipelines/categorize.nf.test

```diff
@@ -0,0 +1,77 @@
+    script "main.nf"
+
+    test("Basic") {
+        tag "pipeline_categorize"
+        tag "pipeline_categorize_basic"
+
+        when {
+            params {
+                input = "$baseDir/tests/data/samplesheets/categorize/basic.csv"
+                outdir = "results"
+            
+                metadata_1_header = "host_scientific_name"
+                metadata_2_header = "host_common_name"
+                metadata_3_header = "food_product"
+                metadata_4_header = "environmental_site"
+                metadata_5_header = "environmental_material"
+
+                transformation = "categorize"
+            }
+        }
+
+        then {
+            assert workflow.success
+            assert path("$launchDir/results").exists()
+
+            // Check Transformation (Machine-Readable)
+            def transformation = path("$launchDir/results/transformation/transformation.csv")
+            assert transformation.exists()
+
+            assert transformation.text.contains("sample,calc_source_type")
+            assert transformation.text.contains("sample1,Human")
+            assert transformation.text.contains("sample2,Animal")
+            assert transformation.text.contains("sample3,Food")
+            assert transformation.text.contains("sample4,Environmental")
+            assert transformation.text.contains("sample5,Unknown")
+            assert transformation.text.contains("sample6,Host Conflict")
+
+            // Check Results (Human-Readable)
+            def results = path("$launchDir/results/transformation/results.csv")
+            assert results.exists()
+
+            assert results.text.contains("sample,sample_name,host_scientific_name,host_common_name,food_product,environmental_site,environmental_material,calc_source_type")
+            assert results.text.contains("sample1,ABC,homo sapiens,human,,,,Human")
+            assert results.text.contains("sample2,DEF,,dog,,,,Animal")
+            assert results.text.contains("sample3,GHI,,,eggs,,,Food")
+            assert results.text.contains("sample4,JKL,,,,farm,wastewater,Environmental")
+            assert results.text.contains("sample5,MNO,,,,,,Unknown")
+            assert results.text.contains("sample6,PQR,homo sapiens,dog,,,,Host Conflict")
+        }
+    }
+
+    test("Missing Headers") {
+        tag "pipeline_categorize"
+        tag "pipeline_categorize_missing_headers"
+
+        when {
+            params {
+                input = "$baseDir/tests/data/samplesheets/categorize/basic.csv"
+                outdir = "results"
+
+                metadata_1_header = "host_scientific_name"
+                metadata_2_header = "host_common_name"
+                metadata_3_header = "food_product"
+
+                transformation = "categorize"
+            }
+        }
+
+        then {
+            assert workflow.failed
+            //TODO: More specific check for if it raised the KeyError with expected message
+        }
```

Comment:
TODO / tests here.

Reply:
This one I need advice on I had tried `assert workflow.errorReport.contains("Expected message")`  here, but it didn't work, even when the python trace was printing the message I was expecting. Not sure which Assertion I'm looking for

Reply:
@ConnorChato could you run this situation and look at what Nextflow outputs? I think it will show that the TRANSFORM_METADATA step failed. You could look at what it outputs and add an assert for that message.

I was thinking about if we wanted to fail in this scenario in a different way (say by the pipeline still succeeding but an error written in an output file). But, the intention with this pipeline is to next create a version that defaults to selecting the relevant metadata fields in IRIDA Next, so this sort of error should be unlikely. Plus, given time constraints, I think any other work on error messages beyond what you have can be pushed for later.